nomad/structs/network.go

```diff
@@ -171,61 +176,131 @@ func (idx *NetworkIndex) Overcommitted() bool {
 	return false
 }
 
-// SetNode is used to setup the available network resources. Returns
-// true if there is a collision
-func (idx *NetworkIndex) SetNode(node *Node) (collide bool, reason string) {
-
-	// COMPAT(0.11): Remove in 0.11
-	// Grab the network resources, handling both new and old
-	var networks []*NetworkResource
+// SetNode is used to initialize a node's network index with available IPs,
+// reserved ports, and other details from a node's configuration and
+// fingerprinting.
+//
+// SetNode must be idempotent as preemption causes SetNode to be called
+// multiple times on the same NetworkIndex, only clearing UsedPorts being
+// calls.
+//
+// An error is returned if the Node cannot produce a consistent NetworkIndex
+// such as if reserved_ports are unparseable. Any such Node is effectively
+// unschedulable for workloads which require networks.
+//
+// Any errors returned by SetNode indicate a bug! The bug may lie in client
+// code not properly validating its configuration or it may lie in improper
+// Node object handling by servers. Users should not be able to cause SetNode
+// to error. Data that cause SetNode to error should be caught upstream such as
+// a client agent refusing to start with an invalid configuration.
+func (idx *NetworkIndex) SetNode(node *Node) error {
+
+	// COMPAT(0.11): Deprecated. taskNetworks are only used for
+	// task.resources.network asks which have been deprecated since before
+	// 0.11.
+	// Grab the network resources, handling both new and old Node layouts
+	// from clients.
+	var taskNetworks []*NetworkResource
 	if node.NodeResources != nil && len(node.NodeResources.Networks) != 0 {
-		networks = node.NodeResources.Networks
+		taskNetworks = node.NodeResources.Networks
 	} else if node.Resources != nil {
-		networks = node.Resources.Networks
+		taskNetworks = node.Resources.Networks
 	}
 
+	// nodeNetworks are used for group.network asks.
 	var nodeNetworks []*NodeNetworkResource
 	if node.NodeResources != nil && len(node.NodeResources.NodeNetworks) != 0 {
 		nodeNetworks = node.NodeResources.NodeNetworks
 	}
```

Comment:
Nitpick: so it doesn't look like we reference or mutate `node.NodeResources` or `nodeNetworks` at all in the two blocks below, so it might be better to move this block down to right above where it's first used (right above the `// TODO: upgrade path?` comment).

Reply:
Good call. Also removing that TODO because I don't even know what it's suggesting.